src/linear_algebra/matrix/circulant.lean

```diff
@@ -0,0 +1,224 @@
+
+@[simp] lemma fin.circulant_inj {v w : fin n → α} :
+  circulant v = circulant w ↔ v = w :=
+(fin.circulant_injective n).eq_iff
+
+lemma transpose_circulant [add_group I] (v : I → α) :
+  (circulant v)ᵀ =  circulant (λ i, v (-i)) :=
+by ext; simp
+
+lemma conj_transpose_circulant [has_star α] [add_group I] (v : I → α) :
+  (circulant v)ᴴ = circulant (star (λ i, v (-i))) :=
+by ext; simp
+
+lemma fin.transpose_circulant : ∀ {n} (v : fin n → α), (circulant v)ᵀ =  circulant (λ i, v (-i))
+| 0     := dec_trivial
+| (n+1) := transpose_circulant
+
+lemma fin.conj_transpose_circulant [has_star α] :
+  ∀ {n} (v : fin n → α), (circulant v)ᴴ = circulant (star (λ i, v (-i)))
+| 0     := dec_trivial
+| (n+1) := conj_transpose_circulant
+
+lemma map_circulant [has_sub I] (v : I → α) (f : α → β) :
+  (circulant v).map f = circulant (λ i, f (v i)) :=
+by ext; simp
+
+lemma circulant_neg [has_neg α] [has_sub I] (v : I → α) :
+  circulant (- v) = - circulant v :=
+by ext; simp
+
+lemma circulant_add [has_add α] [has_sub I] (v w : I → α) :
+  circulant (v + w) = circulant v + circulant w :=
+by ext; simp
+
+lemma circulant_sub [has_sub α] [has_sub I] (v w : I → α) :
+  circulant (v - w) = circulant v - circulant w :=
+by ext; simp
+
+lemma circulant_mul [semiring α] [fintype I] [add_comm_group I] (v w : I → α) :
+  circulant v ⬝ circulant w = circulant (mul_vec (circulant v) w) :=
+begin
+  ext i j,
+  simp only [mul_apply, mul_vec, circulant, dot_product],
+  refine fintype.sum_equiv (equiv.sub_right j) _ _ _,
+  intro x,
+  simp only [equiv.sub_right_apply],
+  congr' 2,
+  abel
+end
+
+lemma fin.circulant_mul [semiring α] :
+  ∀ {n} (v w : fin n → α), circulant v ⬝ circulant w = circulant (mul_vec (circulant v) w)
+| 0     := dec_trivial
+| (n+1) := circulant_mul
+
+/-- circulant matrices commute in multiplication under certain condations. -/
+lemma circulant_mul_comm
+  [comm_semigroup α] [add_comm_monoid α] [fintype I] [add_comm_group I] (v w : I → α) :
+  circulant v ⬝ circulant w = circulant w ⬝ circulant v :=
+begin
+  ext i j,
+  simp only [mul_apply, circulant, mul_comm],
+  refine fintype.sum_equiv ((equiv.sub_left i).trans (equiv.add_right j)) _ _ _,
+  intro x,
+  congr' 2,
+  { simp },
+  { simp only [equiv.coe_add_right, function.comp_app,
+               equiv.coe_trans, equiv.sub_left_apply],
+    abel }
+end
+
+lemma fin.circulant_mul_comm [comm_semigroup α] [add_comm_monoid α] :
+  ∀ {n} (v w : fin n → α), circulant v ⬝ circulant w = circulant w ⬝ circulant v
+| 0     := dec_trivial
+| (n+1) := circulant_mul_comm
+
+/-- `k • circulant v` is another circulantcluant matrix `circulant (k • v)`. -/
+lemma circulant_smul [has_sub I] [has_scalar R α] {k : R} {v : I → α} :
+  circulant (k • v) = k • circulant v :=
+by ext; simp
+
+@[simp] lemma circulant_zero [has_zero α] [has_sub I] :
+  circulant 0 = (0 : matrix I I α) :=
+by ext; simp
+
+@[simp] lemma circulant_single [has_zero α] [has_one α] [decidable_eq I] [add_group I] :
```

Comment:
```suggestion
@[simp] lemma circulant_single_one [has_zero α] [has_one α] [decidable_eq I] [add_group I] :
```
So that you can use this name for the `scalar` lemma below